### PR TITLE
Rename attributes, enable class-level support

### PR DESCRIPTION
**Pull request by Sergio0694**
**Contributes to #8**

This PR renames all source generator attributes for observable properties, and enables class-level support.

## PR Checklist

- [X] Created a feature/dev branch in your fork (vs. submitting directly from a commit on main)
- [X] Based off latest main branch of toolkit
- [X] PR doesn't include merge commits (always rebase on top of our main, if needed)
- [X] Tested code with current [supported SDKs](../#supported)
- [ ] New component
  - [ ] Pull Request has been submitted to the documentation repository [instructions](../blob/main/Contributing.md#docs). Link: 
  - [ ] Added description of major feature to project description for NuGet package (4000 total character limit, so don't push entire description over that)
- [X] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [X] Header has been added to all new source files (run _build/UpdateHeaders.bat_)
- [ ] Contains **NO** breaking changes
- [X] Every new API (including internal ones) has full XML docs
- [X] Code follows all style conventions

### CODE REVIEW

File: tests/CommunityToolkit.Mvvm.UnitTests/Test_ObservablePropertyAttribute.cs [hunk 1]
```diff
@@ -1062,7 +1153,38 @@ public BroadcastingViewModelWithInheritedAttribute(IMessenger messenger)
...
+        {
+        }
+
+        [ObservableProperty]
+        private string? name;
+    }
+
+    [ObservableRecipient]
+    [NotifyRecipients]
+    partial class BroadcastingViewModelWithAttributeAndClassLevelAttribute : ObservableObject
```
Comment (powerdude) [inline tests/CommunityToolkit.Mvvm.UnitTests/Test_ObservablePropertyAttribute.cs:1174]: may want to add a test with VM inheriting from this class too like `BroadcastingViewModelWithInheritedAttributeAndClassLevelAttribute`

File: tests/CommunityToolkit.Mvvm.UnitTests/Test_ObservablePropertyAttribute.cs [hunk 2]
```diff
@@ -1042,15 +1122,26 @@ public BroadcastingViewModel(IMessenger messenger)
         }
 
         [ObservableProperty]
-        [AlsoBroadcastChange]
+        [NotifyRecipients]
+        private string? name;
+    }
+
+    partial class RecipientWithNonBroadcastingProperty : ObservableRecipient
```
Comment (michael-hawker) [inline tests/CommunityToolkit.Mvvm.UnitTests/Test_ObservablePropertyAttribute.cs:1129]: Seems like we have a giant single file for all source generator tests? Should these be broken out into more files? one for base observable property modifies, one for recipients, one for validation at least?

**Review by michael-hawker — CHANGES_REQUESTED** [review]

File: tests/CommunityToolkit.Mvvm.UnitTests/Test_ObservablePropertyAttribute.cs (hunk 1 above)
Comment (Sergio0694) [reply tests/CommunityToolkit.Mvvm.UnitTests/Test_ObservablePropertyAttribute.cs:1194]: Good idea, added in 901d23bb06465d929b3518add458ff540002adfe. Thanks! 😄

File: tests/CommunityToolkit.Mvvm.UnitTests/Test_ObservablePropertyAttribute.cs [hunk 3]
```diff
@@ -1125,10 +1247,10 @@ partial class ModelWithCultureAwarePropertyName
...
     [ObservableRecipient]
-    public sealed partial class ModelWithAlsoBroadcastChangeAndDisplayAttributeLast : ObservableValidator
+    public sealed partial class ModelWithNotifyRecipientsAndDisplayAttributeLast : ObservableValidator
     {
         [ObservableProperty]
-        [AlsoBroadcastChange]
+        [NotifyRecipients]
         [Display(Name = "Foo bar baz")]
         private object? _someProperty;
     }
```
[the remark this answers is not in the document]
Comment (Sergio0694) [reply tests/CommunityToolkit.Mvvm.UnitTests/Test_ObservablePropertyAttribute.cs:1310]: The tests for that are `Test_ObservablePropertyWithValueNamedField_WithValidationAttributesAndValidation` and `Test_ObservablePropertyWithValueNamedField_WithValidationAttributesAndValidation_WithClassLevelAttribute` in this file. Also added another one testing inheritance in 5ef41a9b197335dc19e61607d5db61ca4605c9ce.

File: CommunityToolkit.Mvvm.SourceGenerators/Diagnostics/DiagnosticDescriptors.cs [hunk 4]
```diff
@@ -399,32 +399,64 @@ internal static class DiagnosticDescriptors
...
+    /// </summary>
+    public static readonly DiagnosticDescriptor InvalidTypeForNotifyDataErrorInfoError = new DiagnosticDescriptor(
+        id: "MVVMTK0028",
+        title: "Invalid type for [NotifyDataErrorInfo] attribute",
+        messageFormat: "The type {0} cannot be annotated with [NotifyDataErrorInfo], as it doesn't inherit from ObservableValidator",
+        category: typeof(ObservablePropertyGenerator).FullName,
+        defaultSeverity: DiagnosticSeverity.Error,
+        isEnabledByDefault: true,
+        description: "Types annotated with [NotifyDataErrorInfo] must inherit from ObservableRecipient.",
         helpLinkUri: "https://aka.ms/mvvmtoolkit");
```
[the remark this answers is not in the document]
Comment (Sergio0694) [reply CommunityToolkit.Mvvm.SourceGenerators/Diagnostics/DiagnosticDescriptors.cs:461]: Ah, right. Done 😄

File: CommunityToolkit.Mvvm/ComponentModel/Attributes/NotifyRecipientsAttribute.cs [hunk 5]
```diff
@@ -33,8 +34,12 @@ namespace CommunityToolkit.Mvvm.ComponentModel;
...
 /// </code>
+/// </para>
+/// <para>
+/// This attribute can also be added to a class, and if so it will affect all generated properties in that type and inherited types.
+/// </para>
 /// </summary>
-[AttributeUsage(AttributeTargets.Field, AllowMultiple = false, Inherited = false)]
-public sealed class AlsoBroadcastChangeAttribute : Attribute
+[AttributeUsage(AttributeTargets.Field | AttributeTargets.Class, AllowMultiple = false, Inherited = false)]
+public sealed class NotifyRecipientsAttribute : Attribute
```
Comment (Arlodotexe) [inline CommunityToolkit.Mvvm/ComponentModel/Attributes/NotifyRecipientsAttribute.cs:43]: Are we sure this new name is specific enough to convey that a `PropertyChangedMessage` will be emitted?

Comment (Sergio0694) [reply CommunityToolkit.Mvvm/ComponentModel/Attributes/NotifyRecipientsAttribute.cs:43]: Mmh...
Maybe `[NotifyPropertyChangedRecipients]`? I mean, a bit verbose but seems consistent, and is self describing.

Comment (Arlodotexe) [reply CommunityToolkit.Mvvm/ComponentModel/Attributes/NotifyRecipientsAttribute.cs:43]: `[NotifyPropertyChangedRecipients]` sounds good. It'll cause a lot less questions and confusion.

Comment (Sergio0694) [reply CommunityToolkit.Mvvm/ComponentModel/Attributes/NotifyRecipientsAttribute.cs]: Done in https://github.com/CommunityToolkit/dotnet/pull/288/commits/bcd9f8267a1ca615db009b6ef6b8bc63f74ae28c 👍

**Comment by michael-hawker** [comment]
@Sergio0694 where was the discussion on changing `[ICommand]` again?

**Comment by Sergio0694** [comment]
@michael-hawker I left a comment for tracking here: https://github.com/CommunityToolkit/dotnet/issues/8#issuecomment-1092302948. We also had some discussions on this on Discord and whatnot. Happy to go over things again in API review though if you want 🙂

**Comment by Sergio0694** [comment]
@michael-hawker @Arlodotexe are there any other outstanding issues/questions here, or can we merge this?

**Review by Arlodotexe — APPROVED** [review]

**Pull request merged**